test/e2e/firewall.go

```diff
@@ -0,0 +1,410 @@
+		nodeNames, err := getNonMaterNodeNames(c)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(examineHitNodesFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNames, true)).NotTo(HaveOccurred())
+
+		By("Removing tags from one of the nodes")
+		nodeNamesList := nodeNames.UnsortedList()
+		if len(nodeNamesList) <= 0 {
+			framework.Failf("Expect %v have at least 1 node", nodeNamesList)
+		}
+		removedTags, err := setNodeTags(gceCloud, cloudConfig.ProjectID, cloudConfig.Zone, nodeNamesList[0], []string{})
+		Expect(err).NotTo(HaveOccurred())
+		defer func() {
+			By("Adding tags back to the node")
+			_, err := setNodeTags(gceCloud, cloudConfig.ProjectID, cloudConfig.Zone, nodeNamesList[0], removedTags)
+			Expect(err).NotTo(HaveOccurred())
+			// Make sure the traffic is recover before exit
+			Expect(examineHitNodesFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNames, true)).NotTo(HaveOccurred())
+		}()
+
+		By("Accessing serivce through the external ip and examine got no response from the node without tags")
+		Expect(examineNotHitNodeFromOutside(svcExternalIP, firewallTestHttpPort, firewallTimeoutDefault, nodeNamesList[0], 15)).NotTo(HaveOccurred())
+	})
+})
+
+func createOnlyLocalLoadBalancServiceWithPort(jig *ServiceTestJig, ns, serviceName string, port int32, sourceRanges []string) *api.Service {
+	svc := jig.CreateTCPServiceOrFail(ns, func(svc *api.Service) {
+		svc.Spec.Type = api.ServiceTypeLoadBalancer
+		svc.Spec.Ports = []api.ServicePort{{Port: port}}
+		svc.ObjectMeta.Annotations = map[string]string{
+			service.BetaAnnotationExternalTraffic: service.AnnotationValueExternalTrafficLocal}
+		svc.Spec.LoadBalancerSourceRanges = sourceRanges
+	})
+	svc = jig.WaitForLoadBalancerOrFail(ns, serviceName, loadBalancerCreateTimeoutDefault)
+	jig.SanityCheckService(svc, api.ServiceTypeLoadBalancer)
+	return svc
+}
+
+func examineSvcFirewallRules(gceCloud *gcecloud.GCECloud, svc *api.Service) error {
+	nodesTag, err := getNodesTagFromGroupName(framework.TestContext.CloudConfig.NodeInstanceGroup)
+	if err != nil {
+		return err
+	}
+	framework.Logf("nodesTag: %v", nodesTag)
+
+	fw, err := gceCloud.GetFirewall(makeFirewallName(cloudprovider.GetLoadBalancerName(svc)))
+	if err != nil {
+		return err
+	}
+
+	if len(fw.TargetTags) != 1 {
+		return fmt.Errorf("error number of tags, got %v", fw.TargetTags)
+	}
+	if fw.TargetTags[0] != nodesTag {
+		return fmt.Errorf("expect tag: %v, got: %v", nodesTag, fw.TargetTags[0])
+	}
+	framework.Logf("Tags are correct.")
+
+	if err := verifyProtocolsPorts(fw.Allowed, svc.Spec.Ports); err != nil {
+		return err
+	}
+	framework.Logf("Protocols and ports are correct.")
+
+	if err := verifySourceRanges(fw.SourceRanges, svc.Spec.LoadBalancerSourceRanges); err != nil {
+		return err
+	}
+	framework.Logf("Source ranges are correct.")
+
+	return nil
+}
+
+func makeFirewallName(name string) string {
+	return fmt.Sprintf("k8s-fw-%s", name)
```

Comment:
please add a comment saying that this should match the formatting in cloudprovider/gce/gce.go or wherever, so when it breaks, please will know where to look to find the thing they should be correcting it to.